{% extends "JAR3Doutput/base.html" %}

{% block title %}JAR3D | Home{% endblock %}

{% block page_specific_css_class %}homepage{% endblock page_specific_css_class %}

{% block page_subtitle %}
Scoring RNA loop sequences against known 3D motifs
<a href="{{ BASE }}/main/jar3d-tutorial/" class="btn btn-success pull-right">Tutorial</a>
{% endblock page_subtitle %}

{% block page_content %}
    <div class="row">
        <div class="span8">
            <textarea id='title-input' class="span8" rows='1' tabindex="1"
            placeholder="Optional title for search"></textarea>
            <br>
            <textarea id='input' class="span8" rows='10' tabindex="1"
            placeholder="Single or aligned RNA/DNA sequences or single/multiple loops (fasta-formatted descriptions and secondary structures are optional)">{{input|default:''}}</textarea>
        </div>
        <div class="span4">
            <h4>Sample Input</h4>
            <ul class="examples unstyled">
                <li><a href='#' id="isNoFastaSingleLoop">Single sequence, one loop</a></li>
                <li><a href='#' id="isNoFastaMultipleLoops">Multiple sequences, one loop</a></li>
                <li><a href='#' id="isFastaMultipleLoops">Multiple sequences, one loop (fasta)</a></li>
                <li><a href='#' id="isNoFastaSingleSequenceSS">Single sequence, many loops, with secondary structure</a></li>
                <li><a href='#' id="isFastaSingleSequenceSS">Single sequence, many loops, with secondary structure (fasta)</a></li>
                <li><a href='#' id="isNoFastaMultipleSequencesSS">Multiple sequences, many loops, with secondary structure</a></li>
                <li><a href='#' id="isFastaMultipleSequencesSS">Multiple sequences, many loops, with secondary structure (fasta)</a></li>
            <ul>
        </div>
    </div>

    <div class="row">
        <button class="btn btn-large btn-primary span2" id="submit" tabindex="2">Submit</button>
        <button class="btn btn-danger span1" id="clear">Clear</button>
        &nbsp
        Motif Atlas Version : 
        <select id = "version">
        {% for option in options %}
          <option value={{option}}>{{option}}</option>
        }
        {% endfor %}
        </select>
        <div id="message" class="span3 alert"></div>
    </div>

    <br><br>

    <div class="row">
        <div class="span4">
            <h3>Goal</h3>
            <p>
              <strong>JAR3D</strong> scores RNA hairpin and internal loop sequences against motif groups
              from the <a href="{{BASE}}/rna3dhub/motifs">RNA 3D Motif Atlas</a>, by exact sequence match for sequences already observed in 3D and
              by probabilistic scoring and edit distance for novel sequences.
              RNA hairpin and internal loops are often represented on secondary structure diagrams as if they are unstructured,
              but in fact most are structured by <a href:"http://ndbserver.rutgers.edu/ndbmodule/services/BPCatalog/bpCatalog.html">non-Watson-Crick basepairs</a>, base stacking, and base-backbone interactions.
              Analysis of 3D structures shows that
              different RNA sequences can form the same RNA 3D motif, as is apparent
              in many motif groups in the RNA 3D Motif Atlas.
              JAR3D scores sequences to motif groups based on the ability of the sequences to form the same pattern of interactions observed in 3D structures of the motif.
              Because the RNA 3D Motif Atlas incorporates new RNA 3D structures every four weeks, the performance of JAR3D will improve over time.
              <p>
              Functions of recurrentÂ 3D motifs include:
            </p>
            <ul>
                <li>Architectural roles introducing bends in helices (e.g. kink-turns) or changing helical twist (e.g. C-loops)</li>
                <li>Anchoring RNA tertiary interactions (e.g., GNRA loops and loop-receptors)</li>
                <li>Providing sites for proteins or small molecules to bind.</li>
            </ul>
            <p>
              Inferring the 3D structures of hairpin and internal loops is a step on the way toward
              correctly predicting full RNA 3D structures starting from sequence.
            </p>
        </div>
        <div class="span4">
            <h3>Input and Output</h3>
            <p>
              JAR3D accepts single or multiple sequences having one or many loops.
              See the Examples above.
              <b>One loop:</b>
              To specify the break between strands in internal loops, use an asterisk <b>*</b>.
              Sequence(s) without an asterisk are interpreted as hairpins.
              Internal and hairpin loops should include closing Watson-Crick basepairs, with nucleotides running in 5' to 3' order within each strand.
              Individual loops do not need the nucleotides to be aligned.
              <b>Many loops:</b>
              JAR3D will extract internal and hairpin loops from longer sequences if a dot-bracket secondary structure is provided as the first line
              of the input.
              Multiple sequences need to be aligned to one another.
            </p>
            <p>
              The output shows the best-scoring motif groups from the RNA 3D Motif Atlas using a variety of metrics.
              The user can view a representative
              instance from each motif group and explore the group further at the RNA 3D Motif Atlas page for the motif group.
              The user can also see how their input input sequences align to known 3D instances of a motif.
            </p>
            <h3>Tutorial</h3>
            <p>
            The tutorial can be found <a href="{{ BASE }}/main/jar3d-tutorial/">at this link</a>.
        </div>
        <div class="span4">
            <h3>Method</h3>
              <ol>
                <li>We extract all hairpin and internal loops from a <a href="{{ BASE }}/rna3dhub/nrlist">
              non-redundant set</a> of RNA 3D structures from the PDB/NDB and cluster them in geometrically
            similar <a href="{{ BASE }}/rna3dhub/motifs">families</a>.
              <li>For each recurrent motif, we construct a probabilistic model
              for sequence variability based on a hybrid <strong>Stochastic
              Context-Free Grammar/Markov Random Field</strong> (SCFG/MRF) method we developed.
              <li>To parameterize each model, we use all instances of the motif
              found in the non-redundant dataset and knowledge of RNA nucleotide interactions,
              especially <a href="http://nar.oxfordjournals.org/content/37/7/2294.full">
              isosteric basepairs</a> and their substitution patterns.
              <li>For each motif group, we form an acceptance region that is consistent with the 
              geometry and basepairing of that group. If the score is in the cutoff region, we 
              infer that the new sequence can form the same 3D structure.
              <li>For more infromation you can read:
              <br>
              Identifying novel sequence variants of RNA 3D motifs, by Craig L. Zirbel, James Roll, Blake A. Sweeney, Anton I. Petrov, Meg Pirrung, and Neocles Leontis. Nucl. Acids Res. (2015) doi: 10.1093/nar/gkv651 <a href="http://nar.oxfordjournals.org/content/early/2015/06/29/nar.gkv651.full">link</a>
              </ol>
        </div>
    </div>
{% endblock page_content %}

{% block page_specific_js %}
    <script src="{{ STATIC_URL }}js/jar3dInputValidator.js"></script>
    <script src="{{ STATIC_URL }}js/homepage.js"></script>
{% endblock page_specific_js %}
