{% extends "JAR3Doutput/base.html" %}

{% block title %}JAR3D | Home{% endblock %}

{% block page_specific_css_class %}homepage{% endblock page_specific_css_class %}

{% block page_subtitle %}predicting RNA 3D motifs in sequences{% endblock page_subtitle %}

{% block page_content %}
    <div class="row">
        <div class="span8">
            <textarea id='input' class="span8" rows='10' tabindex="1"
            placeholder="Single or aligned RNA/DNA sequences or single/multiple loops (fasta-formatted descriptions and secondary structures are optional)">{{input|default:''}}</textarea>
        </div>
        <div class="span2">
            <h4>Examples</h4>
            <ul class="examples unstyled">
                <li><a href='#' id="isNoFastaSingleLoop">Single loop</a></li>
                <li><a href='#' id="isFastaSingleLoop">Single loop (fasta)</a></li>
                <li><a href='#' id="isNoFastaMultipleLoops">Multiple loops</a></li>
                <li><a href='#' id="isFastaMultipleLoops">Multiple loops (fasta)</a></li>
                <li><a href='#' id="isNoFastaSingleSequenceNoSS">Single sequence</a></li>
                <li><a href='#' id="isFastaSingleSequenceNoSS">Single sequence (fasta)</a></li>
                <li><a href='#' id="isNoFastaSingleSequenceSS">Single sequence with secondary structure</a></li>
                <li><a href='#' id="isFastaSingleSequenceSS">Single sequence with secondary structure (fasta)</a></li>
            <ul>
        </div>
        <div class="span2">
            <ul class="examples unstyled">
                <li><a href='#' id="isNoFastaMultipleSequencesNoSS">Multiple sequences</a></li>
                <li><a href='#' id="isFastaMultipleSequencesNoSS">Multiple sequences (fasta)</a></li>
                <li><a href='#' id="isNoFastaMultipleSequencesSS">Multiple sequences with secondary structure</a></li>
                <li><a href='#' id="isFastaMultipleSequencesSS">Multiple sequences with secondary structure (fasta)</a></li>
            <ul>
        </div>
    </div>

    <div class="row">
        <button class="btn btn-large btn-primary span2" id="submit" tabindex="2">Submit</button>
        <button class="btn btn-danger span1" id="clear">Clear</button>

        <div id="message" class="span3 alert"></div>
    </div>

    <br><br>

    <div class="row">
        <div class="span4">
            <h3>Goal</h3>
            <p>
              Correctly predicting RNA 3D structures starting from sequence is a major,
              unsolved challenge in bioinformatics. An important sub-goal is the inference
              of the 3D structures of recurrent hairpin and internal RNA 3D motifs
              that appear as unpaired "loops" on secondary structure diagrams.
              RecurrentÂ 3D motif functions include:
            </p>
            <ul>
                <li>Architectural roles introducing bends in helices (e.g. kink-turns) or changing helical twist (e.g. C-loops)</li>
                <li>Anchoring RNA tertiary interactions (e.g., GNRA loops and loop-receptors)</li>
                <li>Providing sites for proteins or small molecules to bind.</li>
            </ul>
            <p>
              Analysis of 3D structures (for example using
              <a href="http://rna.bgsu.edu/webfr3d">WebFR3D</a>) shows that
              different RNA sequences can form the same RNA 3D motif. Since not
              all sequence variants of a given motif are present in the 3D database,
              accurate methods are needed to predict which sequences are likely to form known 3D motifs.
              This is the purpose of <strong>JAR3D</strong>.
            </p>
        </div>
        <div class="span4">
            <h3>Input and Output</h3>
            <p>
              JAR3D recognizes multiple types of input (see Examples above).
              If a secondary structure is provided, it is used to extract
              internal and hairpin loops.
              Unfolded single sequences are folded by <a href="http://mfold.rna.albany.edu/">UNAfold</a>.
              Multiple sequence alignments without secondary structures are folded with
              <a href="http://www.tbi.univie.ac.at/~ivo/RNA/man/RNAalifold.html">RNAalifold</a>.
              To specify chain break in internal loops, use the <code>*</code> symbol.
              Sequence(s) without asterisks and shorter than 25 nucleotides are
              interpreted as hairpins. Internal and hairpin loops should include
              closing Watson-Crick basepairs.
            </p>
            <p>
              The output shows the top scoring motifs from the RNA 3D Motif Atlas.
              The user can view scoring statistics, visualize representative
              motif instances that potentially match input sequences.
            </p>
        </div>
        <div class="span4">
            <h3>Method</h3>
              <ol>
              <li>We extract all hairpin and internal loops from a <a href="http://rna.bgsu.edu/nrlist">
              non-redundant set</a> of RNA 3D structures from the PDB/NDB and cluster them in geometrically
              similar <a href="http://rna.bgsu.edu/rna3dhub/motifs">families</a>.
              <li>For each recurrent motif, we construct a probabilistic model
              for sequence variability based on a hybrid <strong>Stochastic
              Context-Free Grammar/Markov Random Field</strong> (SCFG/MRF) method we developed.
              <li>To parameterize each model, we use all instances of the motif
              found in the non-redundant dataset and RNA knowledge of nucleotide interactions,
              especially <a href="http://nar.oxfordjournals.org/content/37/7/2294.full">
              isosteric basepairs</a> and their substitution patterns.
              <li>Given the sequence of a hairpin or internal loop from a secondary structure as input,
              each SCFG/MRF model calculates the probability that the sequence forms a given 3D motif.
              If the score is in the same range as sequences known to form the 3D structure,
              we infer that the new sequence can form the same 3D structure.
              <br>
              <a href="http://rna.bgsu.edu/main/webapps/jar3d/">Read more</a>
              </ol>
        </div>
    </div>
{% endblock page_content %}

{% block page_specific_js %}
    <script src="{{ STATIC_URL }}js/jar3dInputValidator.js"></script>
    <script src="{{ STATIC_URL }}js/homepage.js"></script>
{% endblock page_specific_js %}

{# Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras rutrum velit in risus luctus id elementum mi sollicitudin. Cras id eros ut sem accumsan pharetra. Mauris sit amet odio rhoncus ipsum ultrices euismod vitae at felis. Vivamus in magna vitae risus sodales vehicula vitae sed dolor. Donec cursus metus a nunc ultricies mollis. Aliquam erat volutpat. Donec eget metus eu turpis luctus malesuada nec eu nibh. Mauris luctus libero in felis pretium tincidunt. Integer nisl augue, rutrum quis tincidunt nec, sagittis sed leo. #}